C++ guidelines for contributing code to EVerest (#2017)
EVerest/everest: docs/source/how-to-guides/c++-coding-guidelines.rst

##################################
C++ Coding Guidelines

> ##################################

Reviewer: This doesn't render into an entry in the left-hand column for me. (I looked into the uploaded doc artifact https://github.com/EVerest/EVerest/actions/runs/23733748283/artifacts/6173375778.)

Perhaps the length of the hash line must align with the length of the title?

This how-to-guide describes best practices for developing C++ code for
EVerest.

The aim is to have a consistent approach in order to avoid coding errors
and aid with code maintenance.

This guide is aimed at new code contributions. Existing code can be updated
but this should be in a separate pull request and not part of functionality
changes.

See :doc:`Contributing to EVerest </project/contributing>` for
more information on how to contribute.

Before merging your contribution please ensure that the merged commit message is
meaningful and describes the changes following review.

Code Layout
===========

C++ code in EVerest is formatted via ``Clang Format`` and this is enforced
via the CI pipelines. This consistent approach is essential to aid with
code reviews ensuring that there are minimal (ideally none) formatting only
changes.

It is recommended to use the EVerest clang format container to reformat your
code before submitting a pull request.

The EVerest clang format container ensures a consistent output using a specific
version of clang format. Unfortunately different versions of clang format
give different results even with the same ``.clang-format`` configuration file.

> The docker container for clang-format is `here:<https://github.com/EVerest/everest-ci>`.

Reviewer 2: ```suggestion
The docker container for clang-format is `here:<https://github.com/EVerest/everest-ci>`_.
```

File Naming
===========

Historically different approaches have been used and it is recommended to be
consistent with the existing code in the repository.

- C++ files should use the extension ``.cpp``
- C++ header files should use the extension ``.hpp``
- C files should use the extension ``.c``
- C header files should use the extension ``.h``
- For EVerest modules file names reflect the class contained within e.g. class
  ``EvseManager`` would have ``EvseManager.cpp`` and ``EvseManager.hpp``
- For library code file names are in snake_case and should reflect the class

>   contained within e.g. class ``OpenSSLProvider`` would have

Reviewer: We probably also need a suggestion for

- the names of subdirectories holding public libraries (in lib/everest/)
- the names of subdirectories holding private libraries (e.g. within modules or other code sections)

  ``openssl_provider.cpp`` and ``openssl_provider.hpp``
- header files and implementation should be co-located in the same directory
- private/internal headers should be in a separate directory (e.g. a sub-directory)
- avoid mixing PascalCase and snake_case in the same filename
- when a directory contains many files consider creating sub-directories to group
  functionality

Naming
======

- classes, and structures, should be named using PascalCase starting with a capital letter
- functions and methods should be in lower case snake_case
- namespaces should relate to the directory hierarchy
- enum values should be consistently named - i.e. avoid mixing PascalCase and
  snake_case
- member variables should be snake_case starting with  `m_`
- function parameters should not match member variables or local variables
- the `this` pointer should only be used when absolutely required

> Example header file called `ocpp/OcppDataModel.hpp`:

Reviewer 2: ```suggestion
Example header file called ``ocpp/OcppDataModel.hpp``:
```

  .. code-block:: cpp

> 

Reviewer 2: ```suggestion
.. code-block:: cpp

```

    // SPDX-License-Identifier: Apache-2.0
    // Copyright 2026 Pionix GmbH and Contributors to EVerest

    /// \file Object to manage the OCPP data model

    #pragma once

    namespace ocpp {

    class OcppDataModel : public OcppDataModelBase {
    private:

>       int _num_connectors{-1};

Reviewer 3: This applies to the rest of the test code, but earlier we said that we use `m_` and here we just prefix `_`

    public:
        OcppDataModel() = default;
        OcppDataModel(int num_connectors) : m_num_connectors(num_connectors) {
        }

        /// \brief check if the object has been configured
        /// \returns true when there is at least one connector

>         bool isConfigured() const {

Reviewer 4: Above is mentioned:

"- functions and methods should be in lower case snake_case"

```suggestion
        bool is_configured() const {
```

        return _num_connectors > 0;
      }

      /// \brief initialise with the count of connectors
      /// \param[in] count - the number of detected connectors
      void initialise(int count);

      operator const OcppDataModelBase&() const {
        return *this;
      }
    };

>     } // namespace ocpp

Reviewer 2: ```suggestion
.. code-block:: cpp

  // SPDX-License-Identifier: Apache-2.0
  // Copyright 2026 Pionix GmbH and Contributors to EVerest

  /// \file Object to manage the OCPP data model

  #pragma once

  namespace ocpp {

  class OcppDataModel : public OcppDataModelBase {
  private:
    int _num_connectors{-1};

  public:
    OcppDataModel() = default;
    OcppDataModel(int num_connectors) : _num_connectors(num_connectors) {
    }

    /// \brief check if the object has been configured
    /// \returns true when there is at least one connector
    bool isConfigured() const {
      return _num_connectors > 0;
    }

    /// \brief initialise with the count of connectors
    /// \param[in] count - the number of detected connectors
    void initialise(int count);

    operator const OcppDataModelBase&() const {
      return *this;
    }
  };

    } // namespace ocpp
```

>     } // namespace ocpp

Reviewer 2: ```suggestion
.. code-block:: cpp

  // SPDX-License-Identifier: Apache-2.0
  // Copyright 2026 Pionix GmbH and Contributors to EVerest

  /// \file Object to manage the OCPP data model

  #pragma once

  namespace ocpp {

  class OcppDataModel : public OcppDataModelBase {
  private:
      int m_num_connectors{-1};

  public:
      OcppDataModel() = default;
      OcppDataModel(int num_connectors) : m_num_connectors(num_connectors) {
      }

      /// \brief check if the object has been configured
      /// \returns true when there is at least one connector
      bool is_configured() const {
          return m_num_connectors > 0;
      }

      /// \brief initialise with the count of connectors
      /// \param[in] count - the number of detected connectors
      void initialise(int count);

      operator const OcppDataModelBase&() const {
          return *this;
      }
  };

  } // namespace ocpp
```

Please remove the surplus indentation as it breaks formatting of the code block a bit.

>     } // namespace ocpp

Reviewer 2: ```suggestion

.. code-block:: cpp

  // SPDX-License-Identifier: Apache-2.0
  // Copyright 2026 Pionix GmbH and Contributors to EVerest

  /// \file Object to manage the OCPP data model

  #pragma once

  namespace ocpp {

  class OcppDataModel : public OcppDataModelBase {
  private:
      int m_num_connectors{-1};

  public:
      OcppDataModel() = default;
      OcppDataModel(int num_connectors) : m_num_connectors(num_connectors) {
      }

      /// \brief check if the object has been configured
      /// \returns true when there is at least one connector
      bool is_configured() const {
          return m_num_connectors > 0;
      }

      /// \brief initialise with the count of connectors
      /// \param[in] count - the number of detected connectors
      void initialise(int count);

      operator const OcppDataModelBase&() const {
          return *this;
      }
  };

  } // namespace ocpp
```

Please remove the surplus indentation as it breaks formatting of the code block a bit.

Example source file called ``ocpp/OcppDataModel.cpp``:

> 
>   .. code-block:: cpp
> 

Reviewer 2: ```suggestion

.. code-block:: cpp

```

> Example source file called `ocpp/OcppDataModel.cpp`:

Reviewer 2: ```suggestion
Example source file called ``ocpp/OcppDataModel.cpp``:
```

    // SPDX-License-Identifier: Apache-2.0
    // Copyright 2026 Pionix GmbH and Contributors to EVerest

    namespace {
    int internal_function(int parameter) {
      return parameter * 4 + 1;
    }

    namespace ocpp {

    void OcppDataModel::initialise(int count) {
      _num_connectors = internal_function(count);
    }

>     } // namespace ocpp

Reviewer 2: ```suggestion
.. code-block:: cpp

  // SPDX-License-Identifier: Apache-2.0
  // Copyright 2026 Pionix GmbH and Contributors to EVerest

  namespace {
  int internal_function(int parameter) {
    return parameter * 4 + 1;
  }

  namespace ocpp {

  void OcppDataModel::initialise(int count) {
    _num_connectors = internal_function(count);
  }

  } // namespace ocpp
```

Header Files
============

> In general inline code should be avoid in header files however C++ templates will

Reviewer 4: ```suggestion
In general inline code should be avoided in header files however C++ templates will
```

often require this and it may be needed for inlining.

> Member variable initialisers should be provided. The following links gives examples

Reviewer 3: ```suggestion
Member variable initializers should be provided. The following links gives examples
```
Just suggesting we keep everything one style of English, since most of it is already American English

of the different types and their uses:

- `Initialisation <https://en.cppreference.com/w/cpp/language/initialization.html>`

> - `Non-static data members <https://en.cppreference.com/w/cpp/language/data_members.html>`

Reviewer 2: ```suggestion
- `Initialisation <https://en.cppreference.com/w/cpp/language/initialization.html>`_
- `Non-static data members <https://en.cppreference.com/w/cpp/language/data_members.html>`_
```

e.g.

```
struct s {
  std::string s3 = "hello";    // 1. copy-initialization
  std::string s4 = {"hello"};  // 2. list-initialization (since C++11)
  std::string s5{'a'};         // 3. list-initialization (since C++11)
};

> ```

Reviewer 2: ```suggestion
.. code-block:: cpp

  struct s {
    std::string s3 = "hello";    // 1. copy-initialization
    std::string s4 = {"hello"};  // 2. list-initialization (since C++11)
    std::string s5{'a'};         // 3. list-initialization (since C++11)
  };
```

> List initialisation is preferred. Be consistent with usage and avoid mixing
> initialisation types (especially 2. and 3.).

Reviewer 3: ```suggestion
List initialization is preferred. Be consistent with usage and avoid mixing
initialization types (especially 2. and 3.).
```

Exceptions
==========

There are always going to be occasions where departure from these guidelines
is warranted. Comments should be used to explain the rationale for the difference.

It is recommended to match the existing style, or create a pull request that
only updates the file to the new standards before making any code changes.
This simplifies reviewing the pull request.

Coding Style
============

> - minimise the use of early returns from a function

Reviewer 3: ```suggestion
- minimize the use of early returns from a function
```

> - minimise the use of early returns from a function

Reviewer: We tend to like early returns. If the "negative" case results in only few lines, or can be considered a "sanity/usefulness check" for the function, it's nice, because it avoid indenting the complete main "positive" block behind its `if` clause.

I would vote against this recommendation.

Author: you might want to read up on MISRA recommendations and why early returns are not good for long term code maintenance.

Also the meaning of should as per https://datatracker.ietf.org/doc/html/rfc2119

Nothing prohibits early returns - it is just not recommended.

an example - initial code that needs some tidy up code adding
```
bool do_what_is_necessary(Args... args){
   if(not args[0]){
       // add tidy-up
       return false;
   } 
   if(not args[1]){
       // add tidy-up
       return false;
   } 
    And many more of those
    ....
    actually do what is necessary
    .....
    // add tidy-up
    return true
}
```

Now has code duplication - a better is to have a common tidy-up section that is easier to maintain:
```
bool do_what_is_necessary(Args... args){
    bool result{false};
    if(arg[0]){
       if(arg[1]){
          if(arg[2]){
             if(and many more of those){
                Do what is necessary
                ....
                result = true;
             }
          }
       }
    } 
    // add tidy-up
    return result;
}
```

Reviewer: Yes, we can agree on this. Let's just keep an eye open to not have spaghetti and large single block indentations when unnecessary.

- try to keep functions short so they can be seen on a single screen

> - always include a ``default:`` clause in a switch statement

Reviewer 3: Could we drop this requirement, I honestly prefer using compile time error for making sure that a new enum is always properly handled than inadvertently defaulting it when it shouldn't be. It helps prevent easy misses.
I suggest dropping to not strictly enforce it

Author: If you enable the correct g++ flags you get that anyway - and I believe it is default.
`-Werror=switch-enum`

test.cpp
```
#include <cstdint>

enum class enum_t : std::uint8_t {
  one,
  two,
  three,
};

int main() {
  enum_t v{enum_t::one};

  switch(v) {
    case enum_t::one:
      break;
    default:
      break;
  }
  return 1;
}
```

compile:
```
james@dell5440:~$ g++ test.cpp 
james@dell5440:~$ g++ test.cpp -Werror=switch-enum
test.cpp: In function ‘int main()’:
test.cpp:12:9: error: enumeration value ‘two’ not handled in switch [-Werror=switch-enum]
   12 |   switch(v) {
      |         ^
test.cpp:12:9: error: enumeration value ‘three’ not handled in switch [-Werror=switch-enum]
cc1plus: some warnings being treated as errors
```

@mlitre 
Can we get this approved now?

Reviewer 3: Was not aware that was the case

- add doxygen comments using ``///`` and ``\brief`` …
- try to use meaningful variable names

> - use `#pragma once` as the include guard

Reviewer 2: ```suggestion
- use ``#pragma once`` as the include guard
```

> - avoid throwing exceptions - keep use of exceptions contained within your class

Reviewer 2: ```suggestion
- always include a ``default:`` clause in a switch statement
- add doxygen comments using ``///`` and ``\brief`` …
- try to use meaningful variable names
- use ``#pragma once`` as the include guard
- avoid throwing exceptions - keep use of exceptions contained within your class
```

> - use `RAII <https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization>` where possible

Reviewer 2: ```suggestion
- use `RAII <https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization>`_ where possible
```

- use existing utility functions (*)

> - minimise external dependencies and ensure a compatible licence exists

Reviewer 3: ```suggestion
- minimize external dependencies and ensure a compatible license exists
```

Author: My preference is English rather than American English and I can't find any statement that American English has to be used.

- design with testing in mind - ensure that your code can be unit tested
- consider object life cycles and the impacts of multiple threads
- add unit tests

* there are a growing set of utility classes and functions in the lib directory.
  Including thread safe queues and object protection (monitor) and bit flags based

>   on enumerations. Where possible use existing implementations and consider generalising new functions

Reviewer 3: ```suggestion
  on enumerations. Where possible use existing implementations and consider generalizing new functions
```

>   for inclusion into the `util`` library.

Reviewer 2: ```suggestion
  for inclusion into the `util` library.
```

> If in doubt have a look at existing code and consider the `Google style guide <https://google.github.io/styleguide/cppguide.html>`

Reviewer 2: ```suggestion
If in doubt have a look at existing code and consider the `Google style guide <https://google.github.io/styleguide/cppguide.html>`_
```

Suggesting a Change
===================

`Zulip <https://lfenergy.zulipchat.com/>` can be used as a starting point to interact with the community about
a change, or enhancement. Once there is some consensus on an approach an issue

> should be raised in the `EVerest github <https://github.com/EVerest/EVerest/issues>`.

Reviewer 2: ```suggestion
`Zulip <https://lfenergy.zulipchat.com/>`_ can be used as a starting point to interact with the community about
a change, or enhancement. Once there is some consensus on an approach an issue
should be raised in the `EVerest github <https://github.com/EVerest/EVerest/issues>`_.
```

See :doc:`Contributing to EVerest </project/contributing>` for
more information on how to contribute.

Issues have a status:

- Backlog - issue needs to be looked at
- Ready - issue is ready for implementation
- In Progress - issue is being worked on
- In review - the updates are in review (e.g. pull request)
- Done - the issue is closed

Please follow the issue template and add information such as the fork/branch
where the issue is being looked at along with any pull request IDs.

It is especially important to check (the comments and status) before starting to

> work on an issue to ensure that someone else hasn't already started to look at it.

Reviewer 2: ```suggestion
work on an issue to ensure that someone else hasn't already started to look at it.

```
